src/main/java/ch/njol/skript/config/Config.java

```diff
@@ -30,285 +15,367 @@
-	public void save(final File f) throws IOException {
+	public void save(File file) throws IOException {
 		separator = defaultSeparator;
-		final PrintWriter w = new PrintWriter(f, "UTF-8");
+		PrintWriter writer = new PrintWriter(file, StandardCharsets.UTF_8);
 		try {
-			main.save(w);
+			main.save(writer);
 		} finally {
-			w.flush();
-			w.close();
+			writer.flush();
+			writer.close();
 		}
 	}
-	
+
 	/**
-	 * Sets this config's values to those in the given config.
-	 * <p>
-	 * Used by Skript to import old settings into the updated config. The return value is used to not modify the config if no new options were added.
-	 *
-	 * @param other
-	 * @return Whether the configs' keys differ, i.e. false == configs only differ in values, not keys.
+	 * @deprecated This copies all values from the other config and sets them in this config,
+	 * which could be destructive for sensitive data if something goes wrong.
+	 * Also removes user comments.
+	 * Use {@link #updateNodes(Config)} instead.
 	 */
+	@Deprecated(forRemoval = true)
 	public boolean setValues(final Config other) {
 		return getMainNode().setValues(other.getMainNode());
 	}
-	
+
+	/**
+	 * @deprecated This copies all values from the other config and sets them in this config,
+	 * which could be destructive for sensitive data if something goes wrong.
+	 * Also removes user comments.
+	 * Use {@link #updateNodes(Config)} instead.
+	 */
+	@Deprecated(forRemoval = true)
 	public boolean setValues(final Config other, final String... excluded) {
 		return getMainNode().setValues(other.getMainNode(), excluded);
 	}
 
 	/**
-	 * Compares the keys and values of this Config and another.
-	 * @param other The other Config.
-	 * @param excluded Keys to exclude from this comparison.
-	 * @return True if there are differences in the keys and their values
-	 *  of this Config and the other Config.
+	 * Updates the nodes of this config with the nodes of another config.
+	 * Used for updating a config file to a newer version.
+	 * <p>
+	 * This method only sets nodes that are missing in this config, thus preserving any existing values.
+	 * </p>
+	 *
+	 * @param newer The newer config to update from.
+	 * @return True if any keys were added to this config, false otherwise.
 	 */
-	public boolean compareValues(Config other, String... excluded) {
-		return getMainNode().compareValues(other.getMainNode(), excluded);
-	}
-	
-	@Nullable
-	public File getFile() {
-		if (file != null) {
-			try {
-				return file.toFile();
-			} catch (Exception e) {
-				return null; // ZipPath, for example, throws undocumented exception
+	public boolean updateNodes(@NotNull Config newer) {
+		Set<Node> newNodes = findNodes(newer.getMainNode());
+		Set<Node> oldNodes = findNodes(getMainNode());
+
+		newNodes.removeAll(oldNodes);
+		Set<Node> nodesToUpdate = new LinkedHashSet<>(newNodes);
+
+		if (nodesToUpdate.isEmpty())
+			return false;
+
+		for (Node node : nodesToUpdate) {
+			Skript.debug("Updating node %s", node);
```

Comment:
Might be worth having a debug message at the beginning like `Updating config <name>`